### PR TITLE
WIP: Middleware: Don't allow invalid licenses

### PR DESCRIPTION
**Pull request by sakjur**
** Problem ** :
When using Grafana Enterprise with a license which has expired we want to disable access to Grafana to prevent undefined behavior from enterprise features suddenly stop working.

** This PR **:
Provides a quick-fix solution based on a middleware (I added it to the auth middleware for now, but could also be a standalone middleware) which blocks invalid but existing licenses. Currently outputs a very basic JSON information text, so that's outstanding before we can consider merging this.

** Notes to reviewer **:
Wanted to bounce an idea of how to handle invalid (but existing) licenses. Should I pursue a middleware based solution for this or do anyone else have a better idea?

### CODE REVIEW

File: pkg/middleware/auth.go [hunk 1]
```diff
@@ -75,6 +75,13 @@ func RoleAuth(roles ...m.RoleType) macaron.Handler {
 
 func Auth(options *AuthOptions) macaron.Handler {
 	return func(c *m.ReqContext) {
+		if options.ReqSignedIn && c.License.HasLicense() && !c.License.HasValidLicense() {
```
Comment (marefr) [inline pkg/middleware/auth.go]: Would add this as a separate licensing middleware. Probably want to execute auth middleware before licensing as well.

**Review by marefr — CHANGES_REQUESTED** [review]

File: pkg/middleware/auth.go (hunk 1 above)
Comment (xlson) [reply pkg/middleware/auth.go]: Yeah, I agree, this should probably be its own middleware.

Comment (sakjur) [reply pkg/middleware/auth.go]: :+1: Is the approach of using middlewares for this (and adding the models.Licensing interface to the request context) reasonable in general?

Comment (marefr) [reply pkg/middleware/auth.go]: Think that's a great idea :+1:

File: pkg/middleware/licensing.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	c.Data["AppSubUrl"] = setting.AppSubUrl
+
+	if exp := c.License.Expiry(); exp != 0 {
+		expDays := int(time.Since(time.Unix(exp, 0)).Hours() / 24)
+		c.Data["ErrorMsg"] = fmt.Sprintf("Your license expired %d days ago.", expDays)
+	} else {
+		c.Data["ErrorMsg"] = "The license is corrupt or cannot be read by Grafana."
+	}
+
+	c.HTML(402, licensingTemplateName)
```
Comment (torkelo) [inline pkg/middleware/licensing.go]: I would not have this here, we do not need a new index (template file), it is very costly to have another. Let's have the middleware only apply to API calls. We can add some props to IndexViewData (index.go) and have the frontend render a license error view instead.

File: pkg/middleware/middleware.go [hunk 3]
```diff
@@ -36,14 +36,15 @@ var (
...
-	ats models.UserTokenService,
+func GetContextHandler(ats models.UserTokenService,
 	remoteCache *remotecache.RemoteCache,
+	license models.Licensing,
 ) macaron.Handler {
 	return func(c *macaron.Context) {
 		ctx := &models.ReqContext{
 			Context:        c,
 			SignedInUser:   &models.SignedInUser{},
+			License:        license,
```
Comment (marefr) [inline pkg/middleware/middleware.go]: Would move updating of ReqContext to the licensing middleware.

Comment (sakjur) [reply pkg/middleware/middleware.go]: Had to take a look to see what you mean. That makes sense, will do that

**Pull request closed without merging**